### PR TITLE
[git registry] add reference field

### PR DESCRIPTION
**Pull request by strega-nil-ms**
This has not yet been tested.

### CODE REVIEW

File: include/vcpkg/registries.h [hunk 1]
```diff
@@ -20,6 +20,53 @@
...
+            auto at = std::find(sv.begin(), sv.end(), '@');
+            if (at == sv.end()) {
+                // old-style, only URL. implicitly HEAD
+                return RepoAndRef{sv.to_string()};
+            }
+            return RepoAndRef{std::string(sv.begin(), at), std::string(at + 1, sv.end())};
+        }
+    };
+
+    inline bool operator==(const RepoAndRef& lhs, const RepoAndRef& rhs) noexcept {
```
Comment (BillyONeal) [inline include/vcpkg/registries.h:49]: Should probably make these hidden friends

File: include/vcpkg/registries.h [hunk 2]
```diff
@@ -20,6 +20,53 @@
 
 namespace vcpkg
 {
+    struct RepoAndRef {
```
Comment (BillyONeal) [inline include/vcpkg/registries.h:23]: Did you intend this to be default constructible?

File: src/vcpkg/registries.cpp [hunk 3]
```diff
@@ -873,6 +873,7 @@ namespace
         constexpr static StringLiteral BASELINE = "baseline";
         constexpr static StringLiteral PATH = "path";
         constexpr static StringLiteral REPO = "repository";
+        constexpr static StringLiteral REFERENCE = "reference";
```
Comment (ras0219-msft) [inline src/vcpkg/registries.cpp:876]: We might want to bikeshed the user-visible name of this field; I feel like there might be something more descriptive of the purpose (`"head"`? `"updates"`?)

File: src/vcpkg/vcpkgpaths.cpp [hunk 4]
```diff
@@ -548,7 +548,7 @@ namespace vcpkg
                         Debug::print("Lockfile value for key '", x.first, "' was not a git commit sha\n");
                         return ret;
                     }
-                    ret.lockdata.emplace(x.first.to_string(), LockFile::EntryData{sv.to_string(), true});
+                    ret.lockdata.emplace(RepoAndRef::parse(x.first), LockFile::EntryData{sv.to_string(), true});
```
Comment (ras0219-msft) [inline src/vcpkg/vcpkgpaths.cpp:551]: I'd prefer to introduce another layer of object than to smoosh the two fields together

**Comment by ras0219-msft** [comment]
Fixed by #155.

**Pull request closed without merging**